File: src/main/java/com/superfriend/superfriend/service/FallbackStrategyService.java
package com.superfriend.superfriend.service;

import com.superfriend.superfriend.dto.McpToolDefinition;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@Slf4j
@Service
public class FallbackStrategyService {

    @Autowired
    @Lazy
    private McpHostService mcpHostService;

    private final Map<String, List<ToolAlternative>> toolAlternatives = new ConcurrentHashMap<>();
    private final Map<String, Integer> failureCounts = new ConcurrentHashMap<>();
    private final Map<String, Long> lastFailureTime = new ConcurrentHashMap<>();

    private static final int MAX_FAILURES_BEFORE_FALLBACK = 2;
    private static final long FAILURE_MEMORY_DURATION_MS = 300000;

    @Data
    public static class ToolAlternative {
        private String toolName;
        private String serverName;
        private double priority;
        private String fallbackCondition;
        private Map<String, String> parameterMapping;
    }

    @Data
    public static class FallbackContext {
        private String originalToolName;
        private String originalServerName;
        private Map<String, Object> originalParameters;
        private String failureReason;
        private String userRequest;
        private List<McpToolDefinition> availableTools;
        private int attemptCount;
    }

    @Data
    public static class FallbackResult {
        private boolean hasAlternative;
        private String alternativeToolName;
        private String alternativeServerName;
        private Map<String, Object> alternativeParameters;
        private String strategy;
        private String reason;
    }

    public FallbackResult findFallback(FallbackContext context) {
        recordFailure(context.getOriginalToolName(), context.getOriginalServerName());

        FallbackResult result = tryDirectAlternative(context);
        if (result.isHasAlternative()) {
            return result;
        }

        result = tryFunctionalAlternative(context);
        if (result.isHasAlternative()) {
            return result;
        }

        result = tryDegradedStrategy(context);
        return result;
    }

    private FallbackResult tryDirectAlternative(FallbackContext context) {
        String toolKey = context.getOriginalServerName() + "__" + context.getOriginalToolName();

        List<ToolAlternative> alternatives = toolAlternatives.get(toolKey);
        if (alternatives != null && !alternatives.isEmpty()) {
            for (ToolAlternative alt : alternatives) {
                if (shouldUseAlternative(alt, context)) {
                    FallbackResult result = new FallbackResult();
                    result.setHasAlternative(true);
                    result.setAlternativeToolName(alt.getToolName());
                    result.setAlternativeServerName(alt.getServerName());
                    result.setAlternativeParameters(mapParameters(context.getOriginalParameters(), alt.getParameterMapping()));
                    result.setStrategy("direct_alternative");
                    result.setReason("使用预设的替代工具: " + alt.getToolName());
                    log.info("找到直接替代工具: {} -> {}", context.getOriginalToolName(), alt.getToolName());
                    return result;
                }
            }
        }

        for (McpToolDefinition tool : context.getAvailableTools()) {
            if (isDirectAlternative(context.getOriginalToolName(), tool.getName())) {
                if (!isToolInFailureState(tool.getName(), tool.getServerName())) {
                    FallbackResult result = new FallbackResult();
                    result.setHasAlternative(true);
                    result.setAlternativeToolName(tool.getName());
                    result.setAlternativeServerName(tool.getServerName());
                    result.setAlternativeParameters(context.getOriginalParameters());
                    result.setStrategy("similar_tool");
                    result.setReason("使用功能相似的工具: " + tool.getName());
                    log.info("找到相似工具: {} -> {}", context.getOriginalToolName(), tool.getName());
                    return result;
                }
            }
        }

        return createNoAlternativeResult();
    }

    private FallbackResult tryFunctionalAlternative(FallbackContext context) {
        String originalTool = context.getOriginalToolName().toLowerCase();
        String userRequest = context.getUserRequest() != null ? context.getUserRequest().toLowerCase() : "";

        if (originalTool.contains("read") || originalTool.contains("get")) {
            return tryReadAlternative(context, userRequest);
        }

        if (originalTool.contains("search") || originalTool.contains("find")) {
            return trySearchAlternative(context, userRequest);
        }

        if (originalTool.contains("write") || originalTool.contains("create")) {
            return tryWriteAlternative(context, userRequest);
        }

        if (originalTool.contains("execute") || originalTool.contains("run")) {
            return tryExecuteAlternative(context, userRequest);
        }

        return createNoAlternativeResult();
    }

    private FallbackResult tryReadAlternative(FallbackContext context, String userRequest) {
        for (McpToolDefinition tool : context.getAvailableTools()) {
            String toolName = tool.getName().toLowerCase();
            if ((toolName.contains("list") || toolName.contains("browse")) 
                    && !isToolInFailureState(tool.getName(), tool.getServerName())) {
                FallbackResult result = new FallbackResult();
                result.setHasAlternative(true);
                result.setAlternativeToolName(tool.getName());
                result.setAlternativeServerName(tool.getServerName());
                result.setAlternativeParameters(context.getOriginalParameters());
                result.setStrategy("list_instead_of_read");
                result.setReason("使用列表工具代替读取工具");
                return result;
            }
        }
        return createNoAlternativeResult();
    }

    private FallbackResult trySearchAlternative(FallbackContext context, String userRequest) {
        for (McpToolDefinition tool : context.getAvailableTools()) {
            String toolName = tool.getName().toLowerCase();
            if ((toolName.contains("read") || toolName.contains("get") || toolName.contains("list"))
                    && !isToolInFailureState(tool.getName(), tool.getServerName())) {
                FallbackResult result = new FallbackResult();
                result.setHasAlternative(true);
                result.setAlternativeToolName(tool.getName());
                result.setAlternativeServerName(tool.getServerName());
                result.setAlternativeParameters(context.getOriginalParameters());
                result.setStrategy("read_instead_of_search");
                result.setReason("使用读取工具代替搜索工具");
                return result;
            }
        }
        return createNoAlternativeResult();
    }

    private FallbackResult tryWriteAlternative(FallbackContext context, String userRequest) {
        for (McpToolDefinition tool : context.getAvailableTools()) {
            String toolName = tool.getName().toLowerCase();
            if (toolName.contains("edit") || toolName.contains("update")
                    && !isToolInFailureState(tool.getName(), tool.getServerName())) {
                FallbackResult result = new FallbackResult();
                result.setHasAlternative(true);
                result.setAlternativeToolName(tool.getName());
                result.setAlternativeServerName(tool.getServerName());
                result.setAlternativeParameters(context.getOriginalParameters());
                result.setStrategy("edit_instead_of_write");
                result.setReason("使用编辑工具代替写入工具");
                return result;
            }
        }
        return createNoAlternativeResult();
    }

    private FallbackResult tryExecuteAlternative(FallbackContext context, String userRequest) {
        for (McpToolDefinition tool : context.getAvailableTools()) {
            String toolName = tool.getName().toLowerCase();
            if (toolName.contains("shell") || toolName.contains("terminal") || toolName.contains("command")
                    && !isToolInFailureState(tool.getName(), tool.getServerName())) {
                FallbackResult result = new FallbackResult();
                result.setHasAlternative(true);
                result.setAlternativeToolName(tool.getName());
                result.setAlternativeServerName(tool.getServerName());
                result.setAlternativeParameters(context.getOriginalParameters());
                result.setStrategy("shell_instead_of_execute");
                result.setReason("使用 Shell 工具代替执行工具");
                return result;
            }
        }
        return createNoAlternativeResult();
    }

    private FallbackResult tryDegradedStrategy(FallbackContext context) {
        if (context.getAttemptCount() >= 3) {
            FallbackResult result = new FallbackResult();
            result.setHasAlternative(false);
            result.setStrategy("abort");
            result.setReason("尝试次数过多，建议放弃并告知用户");
            return result;
        }

        FallbackResult result = new FallbackResult();
        result.setHasAlternative(false);
        result.setStrategy("llm_fallback");
        result.setReason("无可用替代工具，建议让 LLM 直接回答或请求更多信息");
        return result;
    }

    private boolean isDirectAlternative(String originalTool, String candidateTool) {
        if (originalTool.equalsIgnoreCase(candidateTool)) {
            return false;
        }

        String original = originalTool.toLowerCase();
        String candidate = candidateTool.toLowerCase();

        String[] prefixes = {"read", "get", "fetch", "load"};
        if (anyMatch(original, prefixes) && anyMatch(candidate, prefixes)) {
            return true;
        }

        String[] searchPrefixes = {"search", "find", "query", "lookup"};
        if (anyMatch(original, searchPrefixes) && anyMatch(candidate, searchPrefixes)) {
            return true;
        }

        String[] writePrefixes = {"write", "create", "save", "store"};
        if (anyMatch(original, writePrefixes) && anyMatch(candidate, writePrefixes)) {
            return true;
        }

        return false;
    }

    private boolean anyMatch(String text, String[] prefixes) {
        for (String prefix : prefixes) {
            if (text.contains(prefix)) {
                return true;
            }
        }
        return false;
    }

    private boolean shouldUseAlternative(ToolAlternative alt, FallbackContext context) {
        if (alt.getFallbackCondition() == null || alt.getFallbackCondition().isEmpty()) {
            return true;
        }

        String condition = alt.getFallbackCondition().toLowerCase();
        String failureReason = context.getFailureReason() != null ? context.getFailureReason().toLowerCase() : "";

        if (condition.contains("timeout") && failureReason.contains("timeout")) {
            return true;
        }
        if (condition.contains("permission") && failureReason.contains("permission")) {
            return true;
        }
        if (condition.contains("not_found") && (failureReason.contains("not found") || failureReason.contains("不存在"))) {
            return true;
        }

        return false;
    }

    private Map<String, Object> mapParameters(Map<String, Object> originalParams, Map<String, String> mapping) {
        if (mapping == null || mapping.isEmpty()) {
            return originalParams;
        }

        Map<String, Object> mappedParams = new HashMap<>(originalParams);
        for (Map.Entry<String, String> entry : mapping.entrySet()) {
            String oldKey = entry.getKey();
            String newKey = entry.getValue();
            if (originalParams.containsKey(oldKey)) {
                mappedParams.put(newKey, originalParams.get(oldKey));
                if (!oldKey.equals(newKey)) {
                    mappedParams.remove(oldKey);
                }
            }
        }
        return mappedParams;
    }

    private boolean isToolInFailureState(String toolName, String serverName) {
        String key = serverName + "__" + toolName;
        Integer failures = failureCounts.get(key);
        Long lastFailure = lastFailureTime.get(key);

        if (failures == null || lastFailure == null) {
            return false;
        }

        if (System.currentTimeMillis() - lastFailure > FAILURE_MEMORY_DURATION_MS) {
            failureCounts.remove(key);
            lastFailureTime.remove(key);
            return false;
        }

        return failures >= MAX_FAILURES_BEFORE_FALLBACK;
    }

    private void recordFailure(String toolName, String serverName) {
        String key = serverName + "__" + toolName;
        failureCounts.merge(key, 1, Integer::sum);
        lastFailureTime.put(key, System.currentTimeMillis());
        log.debug("记录工具失败: {} (失败次数: {})", key, failureCounts.get(key));
    }

    public void recordSuccess(String toolName, String serverName) {
        String key = serverName + "__" + toolName;
        failureCounts.remove(key);
        lastFailureTime.remove(key);
        log.debug("清除工具失败记录: {}", key);
    }

    public void registerAlternative(String originalTool, String originalServer, ToolAlternative alternative) {
        String key = originalServer + "__" + originalTool;
        toolAlternatives.computeIfAbsent(key, k -> new ArrayList<>()).add(alternative);
        log.info("注册工具替代: {} -> {}", key, alternative.getToolName());
    }

    private FallbackResult createNoAlternativeResult() {
        FallbackResult result = new FallbackResult();
        result.setHasAlternative(false);
        result.setStrategy("none");
        result.setReason("未找到可用的替代方案");
        return result;
    }

    public List<String> getFailedTools() {
        List<String> failed = new ArrayList<>();
        for (Map.Entry<String, Integer> entry : failureCounts.entrySet()) {
            if (entry.getValue() >= MAX_FAILURES_BEFORE_FALLBACK) {
                failed.add(entry.getKey());
            }
        }
        return failed;
    }

    public void clearFailureRecords() {
        failureCounts.clear();
        lastFailureTime.clear();
        log.info("已清除所有工具失败记录");
    }
}

File: superfriend-py/src/superfriend/services/fallback_strategy_service.py
from __future__ import annotations

import time
import structlog
from dataclasses import dataclass, field
from typing import Any

logger = structlog.get_logger()

MAX_FAILURES_BEFORE_FALLBACK = 2
FAILURE_MEMORY_DURATION_MS = 300000


@dataclass
class ToolAlternative:
    tool_name: str = ""
    server_name: str = ""
    priority: float = 1.0
    fallback_condition: str = ""
    parameter_mapping: dict[str, str] = field(default_factory=dict)


@dataclass
class FallbackContext:
    original_tool_name: str = ""
    original_server_name: str = ""
    original_parameters: dict[str, Any] = field(default_factory=dict)
    failure_reason: str = ""
    user_request: str = ""
    available_tools: list[dict[str, Any]] = field(default_factory=list)
    attempt_count: int = 0


@dataclass
class FallbackResult:
    has_alternative: bool = False
    alternative_tool_name: str = ""
    alternative_server_name: str = ""
    alternative_parameters: dict[str, Any] = field(default_factory=dict)
    strategy: str = ""
    reason: str = ""


DEFAULT_ALTERNATIVES: dict[str, list[ToolAlternative]] = {
    "fetch": [
        ToolAlternative(tool_name="puppeteer__navigate", priority=0.8, fallback_condition="javascript_rendering"),
        ToolAlternative(tool_name="web_search", priority=0.6, fallback_condition="content_not_found"),
    ],
    "puppeteer__navigate": [
        ToolAlternative(tool_name="fetch", priority=0.7, fallback_condition="simple_page"),
    ],
    "web_search": [
        ToolAlternative(tool_name="fetch", priority=0.5, fallback_condition="direct_url"),
    ],
    "read_file": [
        ToolAlternative(tool_name="execute_code", priority=0.6, fallback_condition="special_format"),
    ],
    "code_execute": [
        ToolAlternative(tool_name="bash", priority=0.8, fallback_condition="shell_command"),
    ],
}


class FallbackStrategyService:
    def __init__(self):
        self._alternatives: dict[str, list[ToolAlternative]] = dict(DEFAULT_ALTERNATIVES)
        self._failure_counts: dict[str, int] = {}
        self._last_failure_time: dict[str, float] = {}

    def find_fallback(self, context: FallbackContext) -> FallbackResult:
        self._record_failure(context.original_tool_name, context.original_server_name)

        key = f"{context.original_server_name}__{context.original_tool_name}" if context.original_server_name else context.original_tool_name

        if self._failure_counts.get(key, 0) < MAX_FAILURES_BEFORE_FALLBACK:
            return FallbackResult(
                has_alternative=False,
                strategy="retry",
                reason=f"Failures ({self._failure_counts.get(key, 0)}) below threshold ({MAX_FAILURES_BEFORE_FALLBACK}), recommend retry",
            )

        alternatives = self._alternatives.get(context.original_tool_name, [])
        if not alternatives:
            return self._search_similar_tools(context)

        for alt in sorted(alternatives, key=lambda a: a.priority, reverse=True):
            if context.available_tools:
                available_names = {t.get("name", "") for t in context.available_tools}
                if alt.tool_name not in available_names:
                    continue

            mapped_params = self._map_parameters(context.original_parameters, alt.parameter_mapping)
            return FallbackResult(
                has_alternative=True,
                alternative_tool_name=alt.tool_name,
                alternative_server_name=alt.server_name,
                alternative_parameters=mapped_params,
                strategy="alternative_tool",
                reason=f"Switched from {context.original_tool_name} to {alt.tool_name} (condition: {alt.fallback_condition})",
            )

        return self._search_similar_tools(context)

    def _search_similar_tools(self, context: FallbackContext) -> FallbackResult:
        if not context.available_tools:
            return FallbackResult(has_alternative=False, strategy="none", reason="No available tools for fallback")

        original_lower = context.original_tool_name.lower()
        candidates = []
        for tool in context.available_tools:
            tool_name = tool.get("name", "")
            if tool_name == context.original_tool_name:
                continue
            score = 0.0
            if any(kw in tool_name.lower() for kw in original_lower.split("_")):
                score += 2.0
            desc = tool.get("description", "").lower()
            if any(kw in desc for kw in original_lower.split("_")):
                score += 1.0
            if score > 0:
                candidates.append((tool_name, score, tool.get("serverName", "")))

        if candidates:
            candidates.sort(key=lambda x: x[1], reverse=True)
            best_name, _, best_server = candidates[0]
            return FallbackResult(
                has_alternative=True,
                alternative_tool_name=best_name,
                alternative_server_name=best_server,
                strategy="semantic_search",
                reason=f"Found similar tool: {best_name}",
            )

        return FallbackResult(has_alternative=False, strategy="none", reason="No fallback available")

    def _record_failure(self, tool_name: str, server_name: str) -> None:
        key = f"{server_name}__{tool_name}" if server_name else tool_name
        self._failure_counts[key] = self._failure_counts.get(key, 0) + 1
        self._last_failure_time[key] = time.time()

        expired_keys = [
            k for k, t in self._last_failure_time.items()
            if (time.time() - t) * 1000 > FAILURE_MEMORY_DURATION_MS
        ]
        for k in expired_keys:
            self._failure_counts.pop(k, None)
            self._last_failure_time.pop(k, None)

    def reset_failures(self, tool_name: str, server_name: str = "") -> None:
        key = f"{server_name}__{tool_name}" if server_name else tool_name
        self._failure_counts.pop(key, None)
        self._last_failure_time.pop(key, None)

    def record_success(self, tool_name: str, server_name: str = "") -> None:
        self.reset_failures(tool_name, server_name)

    def add_alternative(self, tool_name: str, alternative: ToolAlternative) -> None:
        self._alternatives.setdefault(tool_name, []).append(alternative)

    def _map_parameters(self, original: dict[str, Any], mapping: dict[str, str]) -> dict[str, Any]:
        if not mapping:
            return dict(original)
        result = {}
        for orig_key, new_key in mapping.items():
            if orig_key in original:
                result[new_key] = original[orig_key]
        for k, v in original.items():
            if k not in mapping and k not in result:
                result[k] = v
        return result


fallback_strategy_service = FallbackStrategyService()